### PR TITLE
feat: Telemetry events to posthog

### PR DESCRIPTION
**Pull request by vikramdevtron**
Trigger added to controller on installation and upgrade start, success, failure.

### CODE REVIEW

File: controllers/installer_controller.go [hunk 1]
```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if len(installer.Spec.URL) == 0 {
 		return reconcile.Result{}, fmt.Errorf("url is not specified")
 	}
+	devtronVersion := strings.ReplaceAll(installer.Spec.URL, "https://raw.githubusercontent.com/devtron-labs/devtron/", "")
+	devtronVersion = strings.ReplaceAll(devtronVersion, "/manifests/installation-script", "")
+	if len(devtronVersion) > 8 {
+		devtronVersion = "v1"
```
Comment (pghildiyal) [inline controllers/installer_controller.go:129]: cant understand this logic.. whats the idea? will this same verion be used by orchestrator?

File: controllers/installer_controller.go [hunk 2]
```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		if err != nil {
+			r.Log.Error(err, "failed to get ucid from config map")
+		}
+		if cm != nil && cm.Data != nil && installEvent == -1 {
+			dataMap := cm.Data
+			installEventStr := dataMap["installEvent"]
+			installEvent, err = strconv.Atoi(installEventStr)
+			if err != nil {
+				installEvent = 1 // there might be no key in first time in cm get
+				r.Log.Error(err, "failed to fet install event status from cm")
```
Comment (pghildiyal) [inline controllers/installer_controller.go:174]: fet -> fetch

File: controllers/installer_controller.go [hunk 3]
```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+			}
+		}
 		err = r.Client.Update(context.Background(), installer)
 		if err != nil {
+			if installEvent == 1 {
+				payload = &TelemetryEventDto{UCID: UCID, Timestamp: time.Now(), EventType: InstallationInterrupt, DevtronVersion: devtronVersion}
```
Comment (pghildiyal) [inline controllers/installer_controller.go:180]: UCID may be nil, if line 165 is true

File: controllers/installer_controller.go [hunk 4]
```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		var payload *TelemetryEventDto
+		fromCache := true
+		if installEvent == -1 {
+			fromCache = false
+		}
+		UCID, cm, err := r.getUCID(fromCache)
+		if err != nil {
+			r.Log.Error(err, "failed to get ucid from config map")
+		}
+		if cm != nil && cm.Data != nil && installEvent == -1 {
```
Comment (pghildiyal) [inline controllers/installer_controller.go:168]: If operator restarts then installEvent == -1 but cm.Data != nil and cm.Data.installEvent != nil in that case installEvent will remain -1

File: controllers/installer_controller.go [hunk 5]
```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		}
 		err = r.Client.Update(context.Background(), installer)
 		if err != nil {
+			if installEvent == 1 {
+				payload = &TelemetryEventDto{UCID: UCID, Timestamp: time.Now(), EventType: InstallationInterrupt, DevtronVersion: devtronVersion}
+			} else if installEvent == 2 {
+				payload = &TelemetryEventDto{UCID: UCID, Timestamp: time.Now(), EventType: UpgradeInterrupt, DevtronVersion: devtronVersion}
+			}
+			if installEvent == -1 {
+				payload.EventType = InstallationApplicationError
```
Comment (pghildiyal) [inline controllers/installer_controller.go:185]: If operator restarts then installEvent == -1 but cm.Data != nil and cm.Data.installEvent != nil in that case installEvent will remain -1, but this is not InstallationApplicationError, this is upgrade event

File: controllers/installer_controller.go (hunk 3 above)
Comment (vikramdevtron) [reply controllers/installer_controller.go:180]: in that case event will be publish `InstallationApplicationError`

File: controllers/installer_controller.go (hunk 2 above)
Comment (vikramdevtron) [reply controllers/installer_controller.go]: done

File: controllers/installer_controller.go (hunk 1 above)
Comment (vikramdevtron) [reply controllers/installer_controller.go]: removed devtron version for now.

**Pull request merged**